src/com/sun/jna/ptr/ByReference.java

```diff
@@ -41,4 +42,20 @@
     protected ByReference(int dataSize) {
         setPointer(new Memory(dataSize));
     }
+
+    /**
+     * Express this reference and the value it points to.
+     *
+     * @param value
+     *            The value to display. Callers should pass the result of
+     *            {@code getValue()}.
+     * @return A formatted string containing the object type, reference address, and
+     *         value
+     */
+    protected String toString(Object value) {
```

Comment:
I'm thinking whether it would be a good idea to add a `Object getValue()` method to `ByReference`. The class description requiers all users to override it already and this would be a covariant return so should be a compatible change (should be verified!). If we did that, toString() could be property override and we would not need this more or less static helper method.

What do you think?

Reply:
This was the first direction I tried to go!  Unfortunately it does not work for `IntByReference` and similar that return primitives, the majority of the classes I originally set out to implement.  If we add any sort of `getValue()` interface or abstract method, it breaks the current primitive returns; and any other added method(s) aren't much different than the current implementation.  I tried to choose the option that required the minimal amount of code in the subclasses.

The javadoc does mention "NOTE: this class would ideally be replaced by a generic" which I agree with.  

I suspect most code wouldn't mind the change to a boxed type with an abstract `getValue()` but I don't see any way to do it without breaking compatibility.   Open to other ideas or even deferring this PR to JNA 6.x with boxed return values that won't break most code.

Reply:
The problem with generics is that the "class name" I've relied on isn't available at run time without reflection.  But it'd be fun to just pass a `Ptr<Integer>` around...
```
public abstract class Ptr<V> extends PointerType {

    private final String typeName;

    protected Ptr(int dataSize) {
        setPointer(new Memory(dataSize));
        this.typeName = ((ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[0]
                .getTypeName();
    }

    public abstract void setValue(V value);

    public abstract V getValue();
    
    @Override
    public String toString() {
        V value = getValue();
        if (value == null) {
            return String.format("null@0x%x", Pointer.nativeValue(getPointer()));
        }
        return String.format("%s@0x%x=%s", this.typeName, Pointer.nativeValue(getPointer()), value);
    }
}
```

Reply:
Yeah - with project valhalla (value types on the JVM) we might someday see a common top level type above object, that unifies reference and primitive types.  Maybe then we will also see reified generics. Only drawback for JNA is, that most probably project panama (improved java-native interaction) will also be part of the JDK and the reason for a libffi based solution might go away.

By the way you are right - I overlooked the primitive returns, which will indeed break it.